Improve BackbeatAPI pause/resume logic

benzekrimaha opened the pull request:
Issue: BB-658

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-658 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.7`

* `9.1.0`


Please check the `Fix Version/s` of BB-658, or the target
branch of this pull request.

welansari commented on lib/util/LocationStatusManager.js:
```diff
@@ -238,9 +238,10 @@ class LocationStatusManager {
+     * @returns {undefined}z
      */
     _updateServiceStatusForLocation(location, cb) {
+        const previousState = JSON.parse(JSON.stringify(this._locationStatusStore[location].getValue()));
```
`this._locationStatusStore[location].getValue()` Already returns a JSON no need to stringify and then parse ?

welansari replied:
Actually, this is not really the previous state, if you look at the Mongo query you'll see that it's the value we use to update the document. The previous state can be acquire in the caller of this function.

welansari commented on lib/util/LocationStatusManager.js:
```diff
@@ -252,14 +253,31 @@ class LocationStatusManager {
+                    error: err.message,
+                });
+                
+                const services = Object.keys(previousState);
+                services.forEach(service => {
+                    if (previousState[service] === null) {
+                        return;
+                    }                        
+                    if (previousState[service].paused) {
+                        this._locationStatusStore[location].pauseLocation(service);
+                    } else {
+                        this._locationStatusStore[location].resumeLocation(service, null);
+                    }                            
+                    if (previousState[service].scheduledResume) {
+                        this._locationStatusStore[location].setServiceResumeSchedule(
+                            service, 
+                            new Date(previousState[service].scheduledResume)
+                        );
+                    } else {
+                        this._locationStatusStore[location].setServiceResumeSchedule(service, null);
+                    }  
```
An easier approach would be to initialize the `LocationStatus` class with the value of the older one.

```
// example
const previousState = this._locationStatusStore[location].getValue();
this._locationStatusStore[location] = new LocationStatus(this._supportedServices, previousState);
```

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-658 contains:


* `9.1.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.7`

* `9.1.0`


Please check the `Fix Version/s` of BB-658, or the target
branch of this pull request.

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -438,7 +434,9 @@ class LocationStatusManager {
                     date.setMinutes(date.getMinutes() + 1);
                     this._scheduledResumeJobs[service][location] = schedule.scheduleJob(date,
                         triggerResume.bind(this));
+                    return;
```
should add a test for this case

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -470,9 +468,20 @@ class LocationStatusManager {
-                this._deleteResumeJob(location, service);
-                this._locationStatusStore[location].setServiceResumeSchedule(service, null);
-                return this._updateServiceStatusForLocation(location, next);
+                this._deleteResumeJob(location, service);                
```
here `_deleteResumeJob()` will be applied even if we fail to update the state in mongo : leading to inconsistent state...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -438,7 +434,9 @@ class LocationStatusManager {
                         triggerResume.bind(this));
+                    return;
                 }
+                this._locationStatusStore[location].resumeLocation(service);
```
setting the "in-memory" value afterwards allows to handle errors, but does not actually solve everything: there are still many race conditions here...

- First and foremost, since we set the value afterwards, changing the value quickly will not "toggle" but keep trying to do the same transition
- In addition, the state is changed asynchronously: so we may make multiple changes to mongo in parallel, and the results may not be processed in the same order...and so the (in-mem) state may not match the state in mongo (i.e. send "Pause" to mongo ; send "Resume" to mongo ; process the end of the "Resume" call ; process of the "Suspend" call → in mongo we may have state Resume, but store "suspend" in memory)

→ the way to handle this is to actually have a proper automaton, so we can detect the "conflicting" operations and deal with them (reject, retry, ignore/wait for completion...)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -470,9 +468,20 @@ class LocationStatusManager {
+                this._deleteResumeJob(location, service);                
+                return this._updateServiceStatusForLocation(location, err => {
+                    if (err) {
+                        this._logger.debug('failed to delete scheduled resume', {
```
not tested

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -507,11 +516,23 @@ class LocationStatusManager {
+            if (!paused) {
+                this._updateServiceStatusForLocation(location, err => {
+                    if (err) {
+                        this._logger.debug('failed to pause service', {
```
maybe this log should be moved to `_updateServiceStatusForLocation`, may be less redundant than having it after each call...

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

williamlardier commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+                        stateMachineInstance.scheduledResumeDate.toISOString() : null) !== 
+                    (newScheduledResumeDate ? newScheduledResumeDate.toISOString() : null)) {
+                    this._logger.info('Updating ZK-based state machine from ZK data due to detected changes', 
+                        { key, oldState: stateMachineInstance.state, newState: newSmState, 
+                          oldScheduledResume: stateMachineInstance.scheduledResumeDate, 
+                          newScheduledResume: newScheduledResumeDate,
+                          zkDataUsed: currentZkData });
```
```suggestion
                        {
                            key, oldState: stateMachineInstance.state,
                            newState: newSmState, 
                            oldScheduledResume: stateMachineInstance.scheduledResumeDate, 
                            newScheduledResume: newScheduledResumeDate,
                            zkDataUsed: currentZkData,
                          });
```

williamlardier commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+                }
+
+                if (stateMachineInstance.state !== newSmState || 
+                    (stateMachineInstance.scheduledResumeDate ? 
+                        stateMachineInstance.scheduledResumeDate.toISOString() : null) !== 
```
```suggestion
                    (stateMachineInstance.scheduledResumeDate?.toISOString?.() 
```

Also what if we get two `null` (or `undefined` with my suggestion) -> the condition will pass, is that what we want?

williamlardier commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+            if (this._serviceConfig[service] && !this._serviceConfig[service].isMongo) {
+                this._logger.debug('Initializing state machine for ZK-based service', { key, service, location });
+                try {
+                    sourceData = await this._getZkDataForLocation(service, location);
```
This calls ZK every time we process an operation, is that OK? I saw we increased the timeouts in tests from 1000ms to 2000ms, is this change fine if we pause/resume at high rate?

benzekrimaha replied:
I don't think we can have something such as high rate pause /resume ( we need to wait for the resume to occur before pause , if pause is called twice nothing will happen etc... ) so getting the state here from zk enables us to actally keep the consistency between the states saved  @Kerkesni  what do you think?

williamlardier commented on lib/util/LocationStatusManager.js:
```diff
@@ -12,20 +12,156 @@ const actions = {
+        };
+
+        const transition = transitions[operationType];
+        if (transition) {
+            transition();
+        }
```
```suggestion
            transition?.();
```

+ A log in case there is no transition?

williamlardier reviewed (commented):
I tried to review everything but due to some changes I discuss, some of my comments will likely end up irrelevant, I suggest reading the whole review and work on the "big part"s first, so you don't waste time addressing some of my comments I wrote with the current code

benzekrimaha commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+            if (operation.retryCount < operation.maxRetries) {
+                setTimeout(() => {
+                    this._operationQueue.push(operation);
+                }, 1000 * Math.pow(2, operation.retryCount));
```
we have two retries max for the moment so not really that big

benzekrimaha commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+    /**
+     * Queue an operation for processing
+     */
+    _queueOperation(type, service, locations, options = {}) {
```
For consistency I'd rather stick with how we do it because we'll need to change it everywhere later , we can create a ticket for this if you want

benzekrimaha commented on lib/util/LocationStatusManager.js:
```diff
@@ -12,20 +12,156 @@ const actions = {
+        this.error = null;
+    }
+
+    canTransitionTo(newState) {
```
Went with the `_ `

benzekrimaha commented on lib/util/LocationStatusManager.js:
```diff
@@ -12,20 +12,156 @@ const actions = {
- * @property {string} isMongo - true if service uses mongo to store status
+ * State machine for service location status
  */
+class ServiceLocationStateMachine {
```
Done we do have now a base classe then a ServiceLocationOperation one in separate files , and then ServiceLocationStateMachine with the locationStatusManager extending the base one

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,68 @@
+     * @returns {boolean} True if transition is valid
+     */
+    _canTransitionTo(newState) {
+        const validTransitions = Object.freeze(this.constructor.getTransitions());
```
why double freeze (both here and in LocationStatusManager) ?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,175 @@
+        });
+    }
+
+    static getValidOperations() {
```
I don't understand this method : the valid operations should be the same as the transitions, i.e. in a FSM an operation is valid if there is a transition from the current state using that operation...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,175 @@
+        return OPERATION_TYPES;
+    }
+
+    static getTransitions() {
```
a transition is what happens when an operation is performed on a state.

so if you want to model with array it would more likely be something like: `[state][operation] → { nextState, action}`

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,175 @@
+        return conflicts;
+    }
+
+    _areOperationsConflicting(operation1, operation2) {
```
conflicting operations should not be computed : this should be just a corrolary of the FSM definition (a conflict happens when the operation is not a valid transition from the current state)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,175 @@
+        return this.pendingOperations.delete(operationId);
+    }
+
+    executeStateTransition(operationType, options = {}) {
```
since you introduced a `BaseStateMachine`, the "execution" should be part of that class.

and the action to perform should be part of the "data" defining the FSM (i.e. the map of transitions) : having it here makes the code very hard to follow (and splitting the definition of the FSM), even though we end up hard-coding the FSM in this function....

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                          });
+                }
+                stateMachineInstance.state = newSmState;
+                stateMachineInstance.scheduledResumeDate = newScheduledResumeDate;
```
this updates the state, so must be done through a transition (otherwise no point having a FSM)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                        initialState = sourceData.scheduledResume ?
+                            SERVICE_STATES.SCHEDULED_RESUME : SERVICE_STATES.PAUSED;
+                    } else {
+                        initialState = SERVICE_STATES.RUNNING;
```
this is repeated throughout : either you set a default at the beginning (286), or you set it in every branch, but not both.

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+        const stateMachineInstance = this._stateMachines.get(key);
+
+        if (this._serviceConfig[service] && !this._serviceConfig[service].isMongo) {
+            this._logger.debug('State machine is ZK-based. Refreshing/validating its state from ZK.', 
```
refresh looks the same as init (for ZK backend), should not create so much duplication

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
             lifecycle: {},
         };
+
+        this._stateMachines = new Map();
```
name should be more precise : the purpose of this variables is not to store an arbitrary collection of state machines, but to keep track of the state of each location (and the state is managed through a state machine, but the "data" of a state machine is really just the state of the machine...)

→ the name of the variable should reflect that, so we easily understand what this is about...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                    stateMachineInstance.state = SERVICE_STATES.PAUSED; 
+                    stateMachineInstance.scheduledResumeDate = null;
+                }
+            }
```
should be a transition.
the purpose of the StateMachine is to manage the state, so you must _never_ set the state directly

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                operation.setState(ServiceLocationOperation.STATES.COMPLETED);
+                stateMachine.removePendingOperation(operation.id);
+                return;
+            }
```
all this should be part of the state machine...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+        const stateMachine = await this._getStateMachine(operation.service, operation.location);
+        
+        try {
+            operation.setState(ServiceLocationOperation.STATES.PROCESSING);
```
what is the point of having a "state" on the operation?
an operation is the "event" that triggers a transition in the FSM, so should either it is handled (valid transition & state updated) or rejected (conflict / invalid state) but there is not further processing or state...
(and looking at the PR, I don't see where we use the state of operation)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+
+        this._stateMachines = new Map();
+
+        this._operationQueue = async.queue((operation, callback) => {
```
why do we need a queue, we already have an FSM and node.js is already single-threaded : so in node.js, we can safely make transitions in the FSM wherever we need...
(while a queue opens the risk of queueing similar events which "cancel" each other, but which we will process sequentially...)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+                this._operationQueue.push(operation);
+            });
+            const checkCompletion = () => {
+                const allCompleted = operations.every(op => 
+                    op.state === ServiceLocationOperation.STATES.COMPLETED || 
+                    op.state === ServiceLocationOperation.STATES.CANCELLED
+                );
```
this is reinventing callbacks (async.forEach) or promises (Promise.all) : code will be much simpler and efficient (no timer) if we just trigger each location's transition in parallel...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+     * Execute resume operation
+     */
+    async _executeResumeOperation(location, service, stateMachine) {
+        const paused = this._locationStatusStore[location].getServicePauseStatus(service);
```
should the log be generated automatically from the FSM, to avoid "polluting" the logic here?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+     */
+    async _executePauseOperation(location, service, stateMachine) {
+        const paused = this._locationStatusStore[location].getServicePauseStatus(service);   
+        if (!paused) {
```
is this not the whole purpose of the FSM to handle the state?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+            stateMachine.executeStateTransition(type, options);
+            return;
+        }
+        switch (type) {
```
one of the purpose of FSM is to avoid these kind of switch: if there is some operation to performed, it should be part of the FSM definition (linked to a transition or entry into the state, but there anyway)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -59,11 +203,476 @@ class LocationStatusManager {
+            }
+            await delay(COMPLETION_POLL_INTERVAL_MS);
+        }
+    }
```
should not poll, but use callabcks/promises

benzekrimaha commented:
Note for reviewers : changes on going please do not review yet

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+     */
+    static getStates() {
+        throw new Error('getStates() must be implemented by child class');
+    }
```
not needed

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+     */
+    static getTransitions() {
+        throw new Error('getTransitions() must be implemented by child class');
+    }
```
no reason to use inheritance here : the state machine only needs "data", which can be provided when initializing the state machine (in the constructor).

this will also avoid this weird trick to have inheritance on a static method (using `this.constructor.getTransitions()`)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
+        const stateMachine = await this._getStateMachine(service, location);
+        
+        // Check if operation is valid in current state
+        if (!stateMachine.canExecuteOperation(operationType)) {
```
no reason to do this : this is handled already in `executeOperation` (it will return an error if it fails, and we must then deal with it), and create _opportunities_ for race conditions (i.e. check that we can execute the transition, do something, than perform the transition: but state may have change in between...)

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+     */
+    isInAnyState(states) {
+        return states.includes(this.state);
+    }
```
does not seem to be used, and best avoid this: it can easily lead to anti pattern when the code looks at the state, instead of reacting to automaton callback

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+    canExecuteOperation(operation) {
+        const transitions = this.constructor.getTransitions();
+        return transitions[this.state]?.[operation] !== undefined;
+    }
```
should be removed, the goal of the automaton is to perform kind of "atomic" transitions : which is not possible with such a function... The code using the automaton should simply attempt the transition(s), and it will get an error if the transition fails.

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+     * @param {Object} context - Additional context for the operation
+     * @returns {Object} The transition result { nextState, action }
+     */
+    executeOperation(operation, context = {}) {
```
nit:
```suggestion
    transition(operation, context = {}) {
```

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+        
+        // Execute the action if it's a function
+        if (typeof transition.action === 'function') {
+            transition.action.call(this, context);
```
the callback may be asynchronous : so it must be called `after` updating the state

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
+    /**
+     * Update the in-memory store based on operation
+     */
+    async _updateServiceInStore(service, location, operationType, options = {}) {
```
this is the kind of thing that should be in the automaton callbacks...
it seems you were adding too many abstraction layers, and wrapped everything: on various events, you know exactly which operation you are dealing with, so it may somewhat look like this:

```
if (state.transition(OPERATION.PAUSE)) {
    locationStore.pauseLocation(service);
}
```

or alternatively we simply "issue" transitions, and keep all the code in the callbacks, to be executed when the transition actually happens.

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
+        
+        switch (operationType) {
+            case OPERATION_TYPES.PAUSE:
+                if (!locationStore.getServicePauseStatus(service)) {
```
why do we check the status in the store? Isn't this why we introduced the automaton?

francoisferrand commented on lib/util/BaseStateMachine.js:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Execute an operation and transition to the resulting state
+     * @param {string} operation - The operation to execute
+     * @param {Object} context - Additional context for the operation
```
kind of mixed feelings regarding the `context` passed to callback: the point of having "callbacks" in the automaton is to have the same code executed on transition, and so it should not take `params.... and if we really need params, maybe a better design would be to just "proceed" depending on the success or failure of the transition:

```
if (stateMachine.executeOperation(op)) {
   // do what needs to be done!
} else {
   // failed to transition, maybe return an error?
}
```

maybe this is appropriate, but should double check if this is really the way to go.

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,110 @@
+    }
+
+    static getTransitions() {
+        return Object.freeze({
```
this automaton looks way too simple: the goal of the PR was to handle intermediate states (i.e. pause requested and in-memory state updated, but not applied everywhere), so I would imagine solving it would involve having these intermediate states in here...
(and this is not what fixes the issue, is there really a point refactoring with the state machine?)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -1,31 +1,110 @@
+                        this.metadata.scheduledResumeDate = context.scheduleDate;
+                    }
+                },
+                [OPERATION_TYPES.DELETE_SCHEDULE]: {
```
no difference with the PAUSE operation, is operation really needed?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
+    
+        const stateMachine = this._serviceLocationStates.get(key);
+    
+        // For ZK services, sync with latest state
```
what is the point of _syncing_ the state?
- we are using an automaton so that we can actually manage the state, and there should thus not be anything to "sync" : from the initial state, we should just "listen" to state changes from Zookeeper to update the automaton
- if we reset the state based on zookeeper, then the automaton is not so useful : it should be used to trigger other transition, and thus allow handling the intermediate states (RUNNING → PAUSE_IN_PROGRESS → PAUSED, etc...)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
+                }
+            } catch (err) {
+                this._logger.error('Failed to sync ZK state', { key, error: err.message });
+            }
```
same code as in the beginning of the function (when `!this._serviceLocationStates.has(key)`), should be de-duplicated...

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -52,18 +131,262 @@ class LocationStatusManager {
             lifecycle: {},
         };
+
+        this._serviceLocationStates = new Map();
```
we already store the state (for each workflow/location) in `LocationStatus` instances, in `_locationStatusStore` : so the state is kind of redundant....

francoisferrand reviewed (commented):
Overall, still seems very complex to me, mixing different layers and trying to introduce functions which "abstract everything" and make it very complex to understand what happens.

- We need to find a way to keep the "abstraction" (mongo vs zookeeper) somewhat cleanly separated, and not intermingle it with state management ;
- The state is already managed in memory (that is the starting point of the bug), in LocationStatus, so this  is where the state management would need to be updated ;
- Since we want to fix inconsistent state, it means we need intermediary states (like PAUSE_IN_PROGRESS) in the FSM, and transition both before initiating the backend operation and once it has succeeded (or failed)

---

Since we have multiple processes (or restart), I wonder however if this approach really works in every case : we may write the state to the backend, but then another process could overwrite this state in between, and we would still be inconsistent?
(i.e. maybe the safer -and simpler- way is to rely on the backend information and using more "transactional" operations instead of in-memory cache)

Commit:
remove in-memory pause/resume state

To avoid race conditions and state
inconsistency issues we remove the
in-memory state entirely and replace
it with conditional mongo updates.

Issue: BB-658

Commit:
fix case where hours is set to 0 when scheduling resume

Issue: BB-658

Commit:
fix initializing phase of BackbeatAPI

- Only initialize the DB once (done in the primary process)
- Queue previously scheduled resumes

Issue: BB-658

Commit:
add unit tests

Issue: BB-658

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -136,7 +140,8 @@ class LocationStatusManager {
     _deleteInvalidLocations(locations, cb) {
-        const validLocationNames = Object.keys(validLocations);
+        const validLocationDetails = locations
+            .filter(loc => validLocationNames.includes(loc._id));
```
instead of linear search, should `validLocationNames` be transformed into some sort of object/map, to allow immediately accessing the location?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -136,7 +140,8 @@ class LocationStatusManager {
+        const validLocationDetails = locations
+            .filter(loc => validLocationNames.includes(loc._id));
         const invalidLocations = locations
             .filter(loc => !validLocationNames.includes(loc._id))
```
we traverse and filter the list of locations twice : is there a somewhat "standard" function to "split" the list according to a criteria, in a single call? (maybe split or partition, something like that)

welansari replied:
replaced the two filter functions with a single foreach that builds the two structures

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -263,6 +289,37 @@ class LocationStatusManager {
+        .toArray()
+        .then(docs => {
+            const states = docs.reduce((acc, doc) => {
+                if (doc.value && doc.value[service]) {
```
why do filtering in javascript? should we do this directly in `find()` ?

welansari replied:
It's not filtering i'm transforming the array into a map

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -424,8 +477,20 @@ class LocationStatusManager {
+                        paused: false,
+                        scheduledResume: null,
+                    }
+                }
```
seems a lot to pass on each call to `_updateServiceStatusForLocation`, would be better to abstract the "detail" of mongo implementation/mapping: i.e. passing at the very most the relevant "filters" (`{ paused: true, scheduledResume: date }`) and values (`{ paused: false, scheduledResume: null }`)...

maybe we could make the query something like `{ _id: location, ``[value.${service}]`` : { paused: true, scheduledResume: {$eq: date} } }` (inside _updateServiceStatusForLocation), so everything is consistent?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -424,8 +477,20 @@ class LocationStatusManager {
+            this._updateServiceStatusForLocation(query, update, err => {
                 if (err) {
                     this._logger.error('error resuming scheduled job, retrying in 1min', {
                         method: 'LocationStatusManager._scheduleResumeJob',
```
What happens if the entry is not found?
in that case, it means we basically hit some kind of race condition, and we should ignore the "resume"

--> if it is an error, should not be retried
--> if it is not an error, should not display that location was resumed

francoisferrand replied:
(same for other functions as well)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -192,24 +214,31 @@ class LocationStatusManager {
             }
-            this._logger.info('Locations status setup complete', {
-                method: 'LocationStatusManager._setupLocationStatusStore',
+            if (cluster.isWorker) {
```
not sure why we use worker mode, esp. with only one worker: please create a followup ticket so we can look at it.

welansari replied:
[BB-717](https://scality.atlassian.net/browse/BB-717)

Commit:
return error when conditional update fails

returning an error when the update condition fails and no document
is updated instead of silently failing

Issue: BB-658

Commit:
remove linear filtering from initial setup

Issue: BB-658

Commit:
Abstract MongoDB update logic in LocationStatusManager

Issue: BB-658

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -234,35 +263,74 @@ class LocationStatusManager {
-                }
+    _updateServiceStatusForLocation(location, service, conditions, newStatus, cb) {
+        const query = { _id: location };
+        Object.keys(conditions).forEach(key => {
+            query[`value.${service}.${key}`] = conditions[key];
+        });
```
nit: seems like a `map` call?

welansari replied:
not really

francoisferrand commented on lib/util/LocationStatusManager.js:
we already log error inside `_updateServiceStatusForLocation` : is it worth logging it again here?

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -234,35 +263,74 @@ class LocationStatusManager {
-                    return cb(err);
+            .then(res => {
+                if (res.modifiedCount === 0) {
+                    return cb(invalidStateTransitionError);
```
should log the error here; but maybe not as an "error", just as warning ?
(same for every API)

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -563,15 +631,17 @@ class LocationStatusManager {
-            if (paused) {
-                this._locationStatusStore[location].resumeLocation(service, schedule);
+            const conditions = { paused: true };
+            const newStatus = { paused: !!schedule, scheduledResume: schedule || null };
```
nit: maybe more "readable" with ternary?

```suggestion
            const newStatus = schedule ? { paused: true, scheduledResume: schedule }
                                       : { paused: false, scheduledResume: null };
```

francoisferrand reviewed (commented):
also please rebase 🙏

benzekrimaha reviewed (commented):
can't approve as I opened the PR walid worked on later but lgtm

francoisferrand commented:
/bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/9.1`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/9.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/9.0`



Please check the status of the associated issue BB-658.

Goodbye benzekrimaha.
